src/xarray_regrid/most_common.py

```diff
@@ -0,0 +1,248 @@
+        max_mem: (Approximate) maximum memory in bytes that the regridding routines can
+            use. Note that this is not the total memory consumption and does not include
+            the size of the final dataset.
+            If this kwargs is used, the regridding will be split up into more manageable
+            chunks, and combined for the final dataset.
+
+    Returns:
+        xarray.dataset with regridded categorical data.
+    """
+    da_name = None
+    if isinstance(data, xr.DataArray):
+        da_name = "da" if data.name is None else data.name
+        data = data.to_dataset(name=da_name)
+
+    coords = utils.common_coords(data, target_ds)
+    coord_size = [data[coord].size for coord in coords]
+    mem_usage = np.prod(coord_size) * np.zeros((1,), dtype=np.int64).itemsize
+
+    if max_mem is not None and mem_usage > max_mem:
+        result = split_combine_most_common(
+            data=data, target_ds=target_ds, time_dim=time_dim, max_mem=max_mem
+        )
+    else:
+        result = most_common(data=data, target_ds=target_ds, time_dim=time_dim)
+
+    if da_name is not None:
+        return result[da_name]
+    else:
+        return result
+
+
+def split_combine_most_common(
+    data: xr.Dataset, target_ds: xr.Dataset, time_dim: str, max_mem: int = int(1e9)
+) -> xr.Dataset:
+    """Use a split-combine strategy to reduce the memory use of the most_common regrid.
+
+    Args:
+        data: Input dataset.
+        target_ds: Dataset which coordinates the input dataset should be regrid to.
+        time_dim: Name of the time dimension, as the regridders do not regrid over time.
+            Defaults to "time".
+        max_mem: (Approximate) maximum memory in bytes that the regridding routines can
+            use. Note that this is not the total memory consumption and does not include
+            the size of the final dataset. Defaults to 1e9 (1 GB).
+
+    Returns:
+        xarray.dataset with regridded categorical data.
+    """
+    coords = utils.common_coords(data, target_ds, remove_coord=time_dim)
+    max_datapoints = max_mem // 8  # ~8 bytes per item.
+    max_source_coord_size = max_datapoints ** (1 / len(coords))
+    size_ratios = {
+        coord: (
+            np.median(np.diff(data[coord].to_numpy(), 1))
+            / np.median(np.diff(target_ds[coord].to_numpy(), 1))
+        )
+        for coord in coords
+    }
+    max_coord_size = {
+        coord: int(size_ratios[coord] * max_source_coord_size) for coord in coords
+    }
+
+    blocks = {
+        coord: np.arange(0, target_ds[coord].size, max_coord_size[coord])
+        for coord in coords
+    }
+
+    subsets = []
+    for vals in product(*blocks.values()):
+        isel = {}
+        for coord, val in zip(blocks.keys(), vals, strict=True):
+            isel[coord] = slice(val, val + max_coord_size[coord])
+        subsets.append(most_common(data, target_ds.isel(isel), time_dim=time_dim))
+
+    return xr.merge(subsets)
```

Comment:
For each block, is the redundancy of data points from original dataset close to the boundary of target grid being considered?? (e.g. for the block of target grid with lat=slice(10, 20), then the lat of original data being used should somehow be lat=slice(5, 25), something that covers the target grid, instead of a cut-off).

Reply:
The original dataset is not sliced by the splitter routine. However, in the `most_common` function, a slice is applied to the original dataset. However this covers the target grid + some tolerance.